packages/preview/src/browser/preview-widget.ts

```diff
@@ -0,0 +1,253 @@
+/*
+ * Copyright (C) 2018 TypeFox and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+import {
+    inject,
+    injectable
+} from "inversify";
+import {
+    Resource,
+    DisposableCollection,
+    Disposable
+} from '@theia/core';
+import {
+    BaseWidget,
+    Message,
+    StatefulWidget
+} from '@theia/core/lib/browser';
+import URI from '@theia/core/lib/common/uri';
+import {
+    ResourceProvider,
+    Event,
+    Emitter,
+} from '@theia/core/lib/common';
+import {
+    Workspace,
+    TextDocument,
+    DidChangeTextDocumentParams,
+    Location,
+    Range,
+} from "@theia/languages/lib/common";
+import {
+    PreviewHandler,
+    PreviewHandlerProvider
+} from './preview-handler';
+import { throttle } from 'throttle-debounce';
+
+export const PREVIEW_WIDGET_CLASS = 'theia-preview-widget';
+
+export const PREVIEW_WIDGET_FACTORY_ID = 'preview-widget';
+
+const DEFAULT_ICON = 'fa fa-eye';
+
+@injectable()
+export class PreviewWidget extends BaseWidget implements StatefulWidget {
+
+    protected resource: Resource | undefined;
+    protected previewHandler: PreviewHandler | undefined;
+    protected readonly previewDisposables = new DisposableCollection();
+    protected readonly onDidScrollEmitter = new Emitter<number>();
+    protected readonly onDidDoubleClickEmitter = new Emitter<Location>();
+
+    @inject(ResourceProvider)
+    protected readonly resourceProvider: ResourceProvider;
+
+    @inject(Workspace)
+    protected readonly workspace: Workspace;
+
+    @inject(PreviewHandlerProvider)
+    protected readonly previewHandlerProvider: PreviewHandlerProvider;
+
+    constructor(
+    ) {
+        super();
+        this.id = 'preview';
+        this.title.iconClass = DEFAULT_ICON;
+        this.title.closable = true;
+        this.addClass(PREVIEW_WIDGET_CLASS);
+        this.node.tabIndex = 0;
+        this.startScrollSync();
+        this.startDoubleClickListener();
+        this.update();
+    }
+
+    protected preventScrollNotification: boolean = false;
+    protected startScrollSync(): void {
+        this.node.addEventListener('scroll', throttle(50, (event: UIEvent) => {
```

Comment:
How is `throttle` different to `setTimeout`?

Reply:
`throttle` reduces the frequency of events here. OTOH the previous delay strategy with `setTimeout`/`clearTimeout` caused lagging of sync.